New media_player platform for Russound devices using the RIO protocol

wickerwaka opened the pull request:
media_player platform for Russound devices using the RIO protocol

Auto discovers zones and sources
Handles media metadata from sources that support it
asyncio implementation
Push updates for any zone or source changes so no polling required.

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  platform: russound_rio
  host: 192.168.105.116
  name: Russound
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

wickerwaka committed:
New media_player platform for Russound devices using the RIO protocol
Auto discovers zones and sources
Handles media metadata from sources that support it
asyncio implementation
Push updates for any zone or source changes so no polling required.

wickerwaka committed:
Merge remote-tracking branch 'upstream/master' into russound_rio

wickerwaka committed:
Fixed up linting issues

homeassistant commented:
Hello @wickerwaka,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`9d6913dbbbb1741cc7ecc691e4f10b07af02bf9d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;8448&#x2F;commits&#x2F;9d6913dbbbb1741cc7ecc691e4f10b07af02bf9d) - This commit has something that looks like an email address (mdonlon@riotgames.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hello @wickerwaka,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`7b3997437bd72185d3975f54c1330b9eaf17126a`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;8448&#x2F;commits&#x2F;7b3997437bd72185d3975f54c1330b9eaf17126a) - This commit has something that looks like an email address (mdonlon@riotgames.com). Maybe try linking that to GitHub?.
- [`9d6913dbbbb1741cc7ecc691e4f10b07af02bf9d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;8448&#x2F;commits&#x2F;9d6913dbbbb1741cc7ecc691e4f10b07af02bf9d) - This commit has something that looks like an email address (mdonlon@riotgames.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

Landrash commented:
Documentatio at https://github.com/home-assistant/home-assistant.github.io/pull/2969

wickerwaka commented:
I resolved the CLA issue with this PR, but it doesn't look like the bot removed the label.

joshlawless commented:
Oh, this is good news.  Can't wait to try this out!

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+Each zone is added as a separate media player device. If the source that the
+zone is assigned to is capable of sending metadata such as artist/album name
+then this data will be reported back via the media_* properties.
+"""
```
Please look to other component/platform how that comments need to be.

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+    if host is None or port is None:
+        _LOGGER.error("Invalid config. Expected %s and %s",
+                      CONF_HOST, CONF_PORT)
+        return False
```
Remove this, we do that on PLATFORM_SCHEMA

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+        except CommandException:
+            break
+        if name != '':
+            sources.append((source_id, name))
```
Move that into try block. Better would be if you API can get a list of exists device instead to trow 16 exceptions for that.

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+        else:
+            return None
+
+    def __init__(self, russ, zone_id, name, sources):
```
please move init function on top of class

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+        self._zone_id = zone_id
+        self._sources = sources
+        self._russ.add_zone_callback(self._zone_callback_handler)
+        self._russ.add_source_callback(self._source_callback_handler)
```
Move callback handling into `async_added_to_hass` Interface function

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+    @asyncio.coroutine
+    def async_turn_off(self):
+        """Turn off the zone."""
+        yield from self._russ.send_zone_event(self._zone_id,
```
remove the `@asyncio.coroutine` and return a coroutine

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+    @asyncio.coroutine
+    def async_turn_on(self):
+        """Turn on the zone."""
+        yield from self._russ.send_zone_event(self._zone_id,
```
remove the `@asyncio.coroutine` and return a coroutine

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+    def async_set_volume_level(self, volume):
+        """Set the volume level."""
+        rvol = int(volume * 50.0)
+        yield from self._russ.send_zone_event(self._zone_id,
```
remove the `@asyncio.coroutine` and return a coroutine

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+        for source_id, name in self._sources:
+            if name.lower() != source.lower():
+                continue
+            yield from self._russ.send_zone_event(
```
remove the `@asyncio.coroutine` and return a coroutine

pvizeli commented on homeassistant/components/media_player/russound_rio.py:
```diff
@@ -0,0 +1,267 @@
+                return None
+            return value
+        else:
+            return None
```
make that cleaner with `if xy in (None, "", "---"):`

wickerwaka committed:
Merge branch 'russound_rio' of github.com:wickerwaka/home-assistant into russound_rio

wickerwaka committed:
Addressing PR feedback

Updated russound_rio dependency to 0.1.3
Use enumerate_zones and enumerate_sources methods instead of doing it in
the platform.
Register callbacks in async_added_to_hass coroutine
Corrected behavior of async methods
